### PR TITLE
*: disable insert null to not-null column for single-row insertion in non-strict mode | tidb-test=pr/2410

### LINKED ISSUE #55457: can insert null value into not null column
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

``` sql
set session sql_mode = '';
create table a (id int primary key,b  varchar(10) not null default '');
insert into a values(4, null);
````

### 2. What did you expect to see? (Required)

``` sql
mysql> insert into a values(4, null);
ERROR 1146 (42S02): Table 'test.a' doesn't exist
```

### 3. What did you see instead (Required)

``` sql
mysql> set session sql_mode = '';
Query OK, 0 rows affected (0.00 sec)

mysql> create table a (id int primary key,b  varchar(10) not null default '');
Query OK, 0 rows affected (0.09 sec)

mysql> insert into a values(4, null);
Query OK, 1 row affected, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+---------------------------+
| Level   | Code | Message                   |
+---------+------+---------------------------+
| Warning | 1048 | Column 'b' cannot be null |
+---------+------+---------------------------+
1 row in set (0.00 sec)
```

### 4. What is your TiDB version? (Required)

all version has this issue

``` sql
mysql> select version();
+--------------------+
| version()          |
+--------------------+
| 8.0.11-TiDB-v8.1.0 |
+--------------------+
1 row in set (0.00 sec)
```

### LINKED ISSUE #56381: The phenomenon of executing DML after setting 'sql_mode' to non 'strict_mode' in tidb is not as expected
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
mysql> create table t1(id int primary key,name varchar(20) not null);
Query OK, 0 rows affected (0.05 sec)

mysql> insert into t1 values(1,null);
ERROR 1048 (23000): Column 'name' cannot be null
mysql> 
mysql> set session sql_mode = "";
Query OK, 0 rows affected (0.01 sec)

mysql> insert into t1 values(1,null);
Query OK, 1 row affected, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+------------------------------+
| Level   | Code | Message                      |
+---------+------+------------------------------+
| Warning | 1048 | Column 'name' cannot be null |
+---------+------+------------------------------+
1 row in set (0.00 sec)

mysql> select * from t1;
+----+------+
| id | name |
+----+------+
|  1 |      |
+----+------+
1 row in set (0.01 sec)

### 2. What did you expect to see? (Required)
mysql> insert into t1 values(1,null);
ERROR 1048 (23000): Column 'name' cannot be null

### 3. What did you see instead (Required)
mysql> insert into t1 values(1,null);
Query OK, 1 row affected, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+------------------------------+
| Level   | Code | Message                      |
+---------+------+------------------------------+
| Warning | 1048 | Column 'name' cannot be null |
+---------+------+------------------------------+
1 row in set (0.00 sec)

mysql> select * from t1;
+----+------+
| id | name |
+----+------+
|  1 |      |
+----+------+
1 row in set (0.01 sec)

### 4. What is your TiDB version? (Required)
mysql> select tidb_version();
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                   |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v8.4.0-alpha-290-gf2ed822
Edition: Community
Git Commit Hash: f2ed8228ffc0cee144453ab334cdfe4aa5dac4a5
Git Branch: HEAD
UTC Build Time: 2024-09-27 01:14:04
GoVersion: go1.21.13
Race Enabled: false
Check Table Before Drop: false
Store: tikv |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

### PR DESCRIPTION
**Pull request by joechenrh**
### What problem does this PR solve?

Issue Number: close #55457, close #56381

Problem Summary:

### What changed and how does it work?

Accroding to [MySQL document](https://dev.mysql.com/doc/refman/8.0/en/constraint-invalid-data.html), even in non-strict mode, single null value still can not be inserted into not-null column.

To support this, this PR add a new error group `ErrGroupNoDefault` to detach the handling of default value from the `ErrGroupBadNull`.

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [x] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
- Report an error when inserting null to non-null column in non-strict mode
```

### CODE REVIEW

File: pkg/executor/insert_test.go [hunk 1]
```diff
@@ -592,3 +593,30 @@ func TestInsertLockUnchangedKeys(t *testing.T) {
 		}
 	}
 }
+
+func TestInsertNullInNonStrictMode(t *testing.T) {
```
Comment (bb7133) [inline pkg/executor/insert_test.go:597]: Could you please move(rewrite) this case to ['integration tests'](https://github.com/pingcap/tidb/tree/master/tests/integrationtest)? Thanks.

File: tests/integrationtest/t/executor/insert.test [hunk 2]
```diff
@@ -1620,3 +1620,25 @@ insert into parent values (2, 3) on duplicate key update ref = 3;
...
+create table t2 (id int primary key, col1 varchar(10));
+insert into t2 values (1, null);
+insert ignore into t1 values(5, null);
+set session sql_mode = '';
+-- error 1048
+insert into t1 values(1, null);
+-- error 1048
+insert into t1 set id = 1, col1 = null;
+-- error 1048
+insert t1 VALUES (5, 5) ON DUPLICATE KEY UPDATE col1 = null;
```
Comment (djshow832) [inline tests/integrationtest/t/executor/insert.test:1636]: What about inserting multiple values with `on duplicate key update`? Please add it.

Comment (joechenrh) [reply tests/integrationtest/t/executor/insert.test:1636]: Updated in https://github.com/pingcap/tidb/pull/55477/commits/cc158dd04509cbb77720bca12c96054fc875fb50

**Review by XuHuaiyu — APPROVED** [review]

**Comment by wuhuizuo** [comment]
/hold

please fill the checklist box and release note in the PR description.

**Pull request merged**